comps/llms/text-generation/vllm_ray/vllm_openai.py

```diff
@@ -0,0 +1,183 @@
+        "target_ongoing_requests": 5,
+    },
+    max_ongoing_requests=10,
+)
+@serve.ingress(app)
+class VLLMDeployment:
+    def __init__(
+        self,
+        engine_args: AsyncEngineArgs,
+        response_role: str,
+        lora_modules: Optional[List[LoRAModulePath]] = None,
+        chat_template: Optional[str] = None,
+    ):
+        logger.info(f"Starting with engine args: {engine_args}")
+        self.engine = AsyncLLMEngine.from_engine_args(engine_args)
+
+        # Determine the name of the served model for the OpenAI client.
+        if engine_args.served_model_name is not None:
+            served_model_names = engine_args.served_model_name
+        else:
+            served_model_names = [engine_args.model]
+        self.openai_serving_chat = OpenAIServingChat(
+            self.engine, served_model_names, response_role, lora_modules, chat_template
+        )
+
+    @app.post("/v1/chat/completions")
+    async def create_chat_completion(self, request: ChatCompletionRequest, raw_request: Request):
+        """OpenAI-compatible HTTP endpoint.
+
+        API reference:
+            - https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html
+        """
+        logger.info(f"Request: {request}")
+        generator = await self.openai_serving_chat.create_chat_completion(request, raw_request)
+        if isinstance(generator, ErrorResponse):
+            return JSONResponse(content=generator.model_dump(), status_code=generator.code)
+        if request.stream:
+            return StreamingResponse(content=generator, media_type="text/event-stream")
+        else:
+            assert isinstance(generator, ChatCompletionResponse)
+            return JSONResponse(content=generator.model_dump())
+
+
+def parse_vllm_args(cli_args: Dict[str, str]):
+    """Parses vLLM args based on CLI inputs.
+
+    Currently uses argparse because vLLM doesn't expose Python models for all of the
+    config options we want to support.
+    """
+    parser = make_arg_parser()
+    arg_strings = []
+    for key, value in cli_args.items():
+        arg_strings.extend([f"--{key}", str(value)])
+    logger.info(arg_strings)
+    parsed_args = parser.parse_args(args=arg_strings)
+    return parsed_args
+
+
+def build_app(cli_args: Dict[str, str]) -> serve.Application:
+    """Builds the Serve app based on CLI arguments.
+
+    See https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#command-line-arguments-for-the-server
+    for the complete set of arguments.
+
+    Supported engine arguments: https://docs.vllm.ai/en/latest/models/engine_args.html.
+    """  # noqa: E501
+    if "device" in cli_args.keys():
+        device = cli_args.pop("device")
+    else:
+        try:
+            from habana_frameworks.torch.distributed.hccl import initialize_distributed_hpu
+
+            initialize_distributed_hpu()
+            torch.zeros(1).to("hpu")
+            device = "HPU"
+        except Exception:
+            device = "GPU"
```

Comment:
This guide is for vllm on Guadi. We can remove this line and throw an Error to remind users to install the required environment.

Reply:
removed.